[GEOS-7214] Add scaling support to WPS gs:Download process

ridethepenguin opened the pull request.

simboss commented on src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/DownloadUtilities.java:
```diff
@@ -50,6 +52,10 @@
     /** The LOGGER. */
     private static final Logger LOGGER = Logging.getLogger(DownloadUtilities.class);
 
+    /** Nearest Neighbour interpolation method */
```
We don't need this, there is already an instance available in [ImageUtilities](https://github.com/geotools/geotools/blob/master/modules/library/coverage/src/main/java/org/geotools/resources/image/ImageUtilities.java#L183):

ridethepenguin replied:
Fixed.

simboss commented on src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/ScaleToTarget.java:
```diff
@@ -0,0 +1,416 @@
+     * @param targetSizeY the size of the target image along the Y axis
+     * @throws TransformException
+     */
+    public void setTargetSize(Integer targetSizeX, Integer targetSizeY) throws TransformException {
```
I am not a fun of using Integer VS int, nice style but inefficient.

ridethepenguin replied:
Agreed, but this method will not be called often and using `Integer` allows to deal with missing values more easily.

simboss commented on src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/RasterEstimator.java:
```diff
@@ -126,25 +128,25 @@ public boolean execute(final ProgressListener progressListener, CoverageInfo cov
-            final Rectangle2D originalGridRange = (GridEnvelope2D) reader.getOriginalGridRange();
-            areaRead = originalGridRange.getWidth() * originalGridRange.getHeight();
-
+            scaling = new ScaleToTarget(reader);
```
You are doing an actual read and scale to estimate the size of the final coverage. If by any chance we are not doing a deferred load this might take forever.

simboss commented on src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/ScaleToTarget.java:
```diff
@@ -0,0 +1,416 @@
+ * @author Stefano Costa, GeoSolutions
+ *
+ */
+class ScaleToTarget {
```
It could be interesting to share this code between WPS and WCS 2.0 (since the latter is a core module WPS can lean on it) otherwise we would be duplicating things throughout the code.

ridethepenguin replied:
@simboss  Definitely. However, in order to have this working ASAP, we could leave the refactoring to a second PR. That would also allow us to give it the necessary attention.

dromagnoli commented on src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/DownloadEstimatorProcess.java:
```diff
@@ -1,4 +1,4 @@
-/* (c) 2014 Open Source Geospatial Foundation - all rights reserved
+/* (c) 2015 Open Source Geospatial Foundation - all rights reserved
  * (c) 2001 - 2013 OpenPlans
```
Minor note: I think you should extend the time period by setting 2014 - 2015

dromagnoli commented on src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/DownloadProcess.java:
```diff
@@ -1,4 +1,4 @@
-/* (c) 2014 Open Source Geospatial Foundation - all rights reserved
+/* (c) 2015 Open Source Geospatial Foundation - all rights reserved
  * (c) 2001 - 2013 OpenPlans
```
Same as before

dromagnoli commented on src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/DownloadUtilities.java:
```diff
@@ -1,4 +1,4 @@
-/* (c) 2014 Open Source Geospatial Foundation - all rights reserved
+/* (c) 2015 Open Source Geospatial Foundation - all rights reserved
  * (c) 2001 - 2013 OpenPlans
```
Another one :)

dromagnoli commented on src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/RasterEstimator.java:
```diff
@@ -1,24 +1,22 @@
-/* (c) 2014 Open Source Geospatial Foundation - all rights reserved
+/* (c) 2015 Open Source Geospatial Foundation - all rights reserved
  * (c) 2001 - 2013 OpenPlans
```
Date in header

ridethepenguin commented:
I've addressed feedback and fixed the copyright headers. We also agreed to postpone the WCS refactoring.

Travis build is green: I'm going to merge this soon.
